File: src/renderer/skia/core/SkiaRenderer.ts
import type { CanvasKit, Canvas, Surface } from 'canvaskit-wasm';
import type { RenderElement, RenderOptions, Transform } from '../../types';

export class SkiaRenderer {
  private ck: CanvasKit;
  private surface: Surface | null;
  private canvas: Canvas;
  private fontManager: any;
  private elementCache = new Map<string, any>();
  private imageCache = new Map<string, any>();

  constructor(canvasKit: CanvasKit, canvasElement: HTMLCanvasElement) {
    this.ck = canvasKit;

    // 优先使用 WebGL，回退到 CPU
    this.surface = canvasKit.MakeWebGLCanvasSurface(canvasElement);
    if (!this.surface) {
      console.warn('WebGL not available, falling back to CPU rendering');
      this.surface = canvasKit.MakeSWCanvasSurface(canvasElement);
    }

    if (!this.surface) {
      throw new Error('Failed to create Skia surface');
    }

    this.canvas = this.surface.getCanvas();
    // @ts-ignore - FontMgr API variations
    this.fontManager = canvasKit.FontMgr?.RefDefault?.() || canvasKit.FontMgr?.RefEmpty?.();
  }

  // 主渲染函数
  render(elements: RenderElement[], options: RenderOptions = {}) {
    // 清空画布
    const clearColor = this.parseColor(options.background || '#ffffff');
    this.canvas.clear(clearColor);

    // 应用视口变换
    if (options.viewport) {
      this.applyViewport(options.viewport);
    }

    // 渲染网格（设计模式）
    if (options.showGrid) {
      this.renderGrid(options.gridSize || 20);
    }

    // 渲染所有元素
    for (const element of elements) {
      if (element.visible) {
        this.renderElement(element);
      }
    }

    // 渲染叠加层（用于设计模式）
    if (options.overlays) {
      this.renderOverlays(options.overlays);
    }

    // 刷新到屏幕
    this.surface?.flush();
  }

  private renderElement(element: RenderElement) {
    this.canvas.save();

    // 应用变换
    if (element.transform) {
      this.applyTransform(element.transform);
    }

    // 应用样式
    if (element.style.opacity !== undefined && element.style.opacity < 1) {
      // @ts-ignore - saveLayerAlpha may not be in type definitions
      this.canvas.saveLayerAlpha?.(null, element.style.opacity * 255) || this.canvas.saveLayer(null, this.createOpacityPaint(element.style.opacity));
    }

    // 根据类型渲染
    switch (element.type) {
      case 'text':
        this.renderText(element);
        break;
      case 'rect':
        this.renderRect(element);
        break;
      case 'circle':
        this.renderCircle(element);
        break;
      case 'path':
        this.renderPath(element);
        break;
      case 'image':
        this.renderImage(element);
        break;
      case 'group':
        this.renderGroup(element);
        break;
    }

    if (element.style.opacity !== undefined && element.style.opacity < 1) {
      this.canvas.restore(); // restore layer
    }

    this.canvas.restore();
  }

  private renderText(element: RenderElement) {
    const { content, fontSize = 14, fontFamily = 'Arial', color = '#000000' } = element.data;

    if (!content) return;

    const paint = new this.ck.Paint();
    paint.setColor(this.parseColor(color));
    paint.setAntiAlias(true);

    // 创建字体
    // @ts-ignore - FontStyle API variations
    const typeface = this.fontManager.matchFamilyStyle(fontFamily, this.ck.FontStyle?.Normal || 0);
    const font = new this.ck.Font(typeface, fontSize);

    // 简单文本渲染
    if (!content.includes('\n')) {
      this.canvas.drawText(content, 0, fontSize, paint, font);
    } else {
      // 多行文本
      const lines = content.split('\n');
      let y = fontSize;
      for (const line of lines) {
        this.canvas.drawText(line, 0, y, paint, font);
        y += fontSize * 1.2; // 行高
      }
    }

    paint.delete();
    font.delete();
  }

  private renderRect(element: RenderElement) {
    const paint = new this.ck.Paint();
    paint.setAntiAlias(true);

    const rect = this.ck.LTRBRect(
      0, 0,
      element.style.width || 100,
      element.style.height || 100
    );

    // 填充
    if (element.style.fill) {
      paint.setStyle(this.ck.PaintStyle.Fill);
      paint.setColor(this.parseColor(element.style.fill));
      this.canvas.drawRect(rect, paint);
    }

    // 描边
    if (element.style.stroke) {
      paint.setStyle(this.ck.PaintStyle.Stroke);
      paint.setColor(this.parseColor(element.style.stroke));
      paint.setStrokeWidth(element.style.strokeWidth || 1);
      this.canvas.drawRect(rect, paint);
    }

    paint.delete();
  }

  private renderCircle(element: RenderElement) {
    const paint = new this.ck.Paint();
    paint.setAntiAlias(true);

    const radius = Math.min(
      (element.style.width || 100) / 2,
      (element.style.height || 100) / 2
    );

    const cx = (element.style.width || 100) / 2;
    const cy = (element.style.height || 100) / 2;

    // 填充
    if (element.style.fill) {
      paint.setStyle(this.ck.PaintStyle.Fill);
      paint.setColor(this.parseColor(element.style.fill));
      this.canvas.drawCircle(cx, cy, radius, paint);
    }

    // 描边
    if (element.style.stroke) {
      paint.setStyle(this.ck.PaintStyle.Stroke);
      paint.setColor(this.parseColor(element.style.stroke));
      paint.setStrokeWidth(element.style.strokeWidth || 1);
      this.canvas.drawCircle(cx, cy, radius, paint);
    }

    paint.delete();
  }

  private renderPath(element: RenderElement) {
    const path = new this.ck.Path();
    const commands = element.data.commands || [];

    for (const cmd of commands) {
      switch (cmd.type) {
        case 'M':
          path.moveTo(cmd.x, cmd.y);
          break;
        case 'L':
          path.lineTo(cmd.x, cmd.y);
          break;
        case 'Q':
          path.quadTo(cmd.x1, cmd.y1, cmd.x, cmd.y);
          break;
        case 'C':
          path.cubicTo(cmd.x1, cmd.y1, cmd.x2, cmd.y2, cmd.x, cmd.y);
          break;
        case 'Z':
          path.close();
          break;
      }
    }

    const paint = new this.ck.Paint();
    paint.setAntiAlias(true);

    if (element.style.fill) {
      paint.setStyle(this.ck.PaintStyle.Fill);
      paint.setColor(this.parseColor(element.style.fill));
      this.canvas.drawPath(path, paint);
    }

    if (element.style.stroke) {
      paint.setStyle(this.ck.PaintStyle.Stroke);
      paint.setColor(this.parseColor(element.style.stroke));
      paint.setStrokeWidth(element.style.strokeWidth || 1);
      this.canvas.drawPath(path, paint);
    }

    path.delete();
    paint.delete();
  }

  private async renderImage(element: RenderElement) {
    const { src } = element.data;
    if (!src) return;

    let image = this.imageCache.get(src);

    if (!image) {
      try {
        const response = await fetch(src);
        const arrayBuffer = await response.arrayBuffer();
        image = this.ck.MakeImageFromEncoded(new Uint8Array(arrayBuffer));
        if (image) {
          this.imageCache.set(src, image);
        }
      } catch (error) {
        console.error('Failed to load image:', src, error);
        return;
      }
    }

    if (!image) return;

    const paint = new this.ck.Paint();
    const srcRect = this.ck.LTRBRect(0, 0, image.width(), image.height());
    const destRect = this.ck.LTRBRect(
      0, 0,
      element.style.width || 100,
      element.style.height || 100
    );

    this.canvas.drawImageRect(image, srcRect, destRect, paint);
    paint.delete();
  }

  private renderGroup(element: RenderElement) {
    if (!element.children) return;

    for (const child of element.children) {
      if (child.visible) {
        this.renderElement(child);
      }
    }
  }

  // 渲染网格背景
  private renderGrid(gridSize: number) {
    const paint = new this.ck.Paint();
    paint.setStyle(this.ck.PaintStyle.Stroke);
    paint.setColor(this.parseColor('#e0e0e0'));
    paint.setStrokeWidth(0.5);
    paint.setAntiAlias(true);

    const width = this.surface?.width() || 1200;
    const height = this.surface?.height() || 800;

    // 垂直线
    for (let x = 0; x <= width; x += gridSize) {
      const path = new this.ck.Path();
      path.moveTo(x, 0);
      path.lineTo(x, height);
      this.canvas.drawPath(path, paint);
      path.delete();
    }

    // 水平线
    for (let y = 0; y <= height; y += gridSize) {
      const path = new this.ck.Path();
      path.moveTo(0, y);
      path.lineTo(width, y);
      this.canvas.drawPath(path, paint);
      path.delete();
    }

    paint.delete();
  }

  // 渲染叠加层（设计模式用）
  private renderOverlays(overlays: any[]) {
    for (const overlay of overlays) {
      switch (overlay.type) {
        case 'selection':
          this.renderSelectionOverlay(overlay);
          break;
        case 'grid':
          this.renderGridOverlay(overlay);
          break;
        case 'ruler':
          this.renderRulerOverlay(overlay);
          break;
      }
    }
  }

  private renderSelectionOverlay(overlay: any) {
    const paint = new this.ck.Paint();
    paint.setStyle(this.ck.PaintStyle.Stroke);
    paint.setColor(this.parseColor('#0080ff'));
    paint.setStrokeWidth(2);

    // 虚线效果
    const dashEffect = this.ck.PathEffect.MakeDash([5, 5], 0);
    paint.setPathEffect(dashEffect);

    const rect = this.ck.LTRBRect(
      overlay.data.x,
      overlay.data.y,
      overlay.data.x + overlay.data.width,
      overlay.data.y + overlay.data.height
    );

    this.canvas.drawRect(rect, paint);

    // 绘制控制点
    paint.setPathEffect(null);
    paint.setStyle(this.ck.PaintStyle.Fill);
    paint.setColor(this.parseColor('#ffffff'));

    const handles = this.getResizeHandles(overlay.data);
    for (const handle of handles) {
      this.canvas.drawCircle(handle.x, handle.y, 4, paint);
    }

    paint.delete();
  }

  private renderGridOverlay(overlay: any) {
    const paint = new this.ck.Paint();
    paint.setStyle(this.ck.PaintStyle.Stroke);
    paint.setColor(this.ck.Color4f(0, 0, 0, 0.1));
    paint.setStrokeWidth(0.5);

    const { width, height, gridSize = 10 } = overlay.data;

    // 垂直线
    for (let x = 0; x <= width; x += gridSize) {
      this.canvas.drawLine(x, 0, x, height, paint);
    }

    // 水平线
    for (let y = 0; y <= height; y += gridSize) {
      this.canvas.drawLine(0, y, width, y, paint);
    }

    paint.delete();
  }

  private renderRulerOverlay(_overlay: any) {
    // 标尺渲染逻辑
  }

  // 工具函数
  private parseColor(color: string): Float32Array {
    if (color.startsWith('#')) {
      const hex = color.replace('#', '');
      const r = parseInt(hex.substr(0, 2), 16) / 255;
      const g = parseInt(hex.substr(2, 2), 16) / 255;
      const b = parseInt(hex.substr(4, 2), 16) / 255;
      const a = hex.length > 6 ? parseInt(hex.substr(6, 2), 16) / 255 : 1;
      return this.ck.Color4f(r, g, b, a);
    }
    // 默认黑色
    return this.ck.Color4f(0, 0, 0, 1);
  }

  // 创建透明度画笔
  private createOpacityPaint(opacity: number) {
    const paint = new this.ck.Paint();
    paint.setAlphaf(opacity);
    return paint;
  }

  private applyTransform(transform: Transform) {
    if (transform.translate) {
      this.canvas.translate(transform.translate.x, transform.translate.y);
    }
    if (transform.scale) {
      this.canvas.scale(transform.scale.x, transform.scale.y);
    }
    if (transform.rotate) {
      const origin = transform.origin || { x: 0, y: 0 };
      this.canvas.rotate(transform.rotate, origin.x, origin.y);
    }
  }

  private applyViewport(viewport: any) {
    this.canvas.save();
    this.canvas.translate(-viewport.x, -viewport.y);
    this.canvas.scale(viewport.zoom, viewport.zoom);
  }

  private getResizeHandles(bounds: any) {
    const { x, y, width, height } = bounds;
    return [
      { x: x, y: y },                          // 左上
      { x: x + width / 2, y: y },              // 中上
      { x: x + width, y: y },                  // 右上
      { x: x + width, y: y + height / 2 },     // 右中
      { x: x + width, y: y + height },         // 右下
      { x: x + width / 2, y: y + height },     // 中下
      { x: x, y: y + height },                 // 左下
      { x: x, y: y + height / 2 },             // 左中
    ];
  }

  // 导出功能
  async exportToPNG(): Promise<Uint8Array | null> {
    if (!this.surface) return null;
    const image = this.surface.makeImageSnapshot();
    const bytes = image.encodeToBytes();
    image.delete();
    return bytes;
  }

  // 清理资源
  dispose() {
    // 清理缓存
    this.elementCache.clear();
    this.imageCache.forEach(image => image?.delete());
    this.imageCache.clear();

    // 清理 surface
    this.surface?.delete();
  }
}

File: src-tauri/src/renderer/core/skia_renderer.rs
use skia_safe::{
    surfaces, pdf, svg, scalar,
    Surface, Canvas, Paint, PaintStyle, Path, Font, FontMgr, FontStyle,
    Color, Color4f, Data, Image, ImageInfo, ColorType, AlphaType,
    Rect, Point as SkPoint, Size, Matrix, PathEffect,
    EncodedImageFormat, TextBlob, Typeface, Document
};
use crate::renderer::types::{
    RenderElement, RenderOptions, Transform, ElementType, ElementStyle,
    Shadow, BlendMode, ExportOptions, Viewport, Overlay, RenderQuality,
    Point
};
use anyhow::{Result, anyhow};
use std::collections::HashMap;

pub struct SkiaRenderer {
    surface: Surface,
    width: i32,
    height: i32,
    font_mgr: FontMgr,
    image_cache: HashMap<String, Data>,
}

impl SkiaRenderer {
    pub fn new(width: i32, height: i32) -> Result<Self> {
        let surface = surfaces::raster_n32_premul((width, height))
            .ok_or_else(|| anyhow!("Failed to create surface"))?;

        Ok(Self {
            surface,
            width,
            height,
            font_mgr: FontMgr::new(),
            image_cache: HashMap::new(),
        })
    }

    /// 主渲染函数
    pub fn render(&mut self, elements: &[RenderElement]) -> Result<()> {
        self.render_with_options(elements, None)
    }

    /// 带选项的渲染函数
    pub fn render_with_options(&mut self, elements: &[RenderElement], options: Option<&RenderOptions>) -> Result<()> {
        // 先解析背景色，避免borrow冲突
        let bg_color = options
            .and_then(|o| o.background.as_ref())
            .map(|c| self.parse_color(c))
            .unwrap_or(Color::WHITE);

        // 获取canvas并设置背景
        let canvas = self.surface.canvas();
        canvas.clear(bg_color);

        // 渲染网格（如果启用）
        if let Some(opts) = options {
            if opts.show_grid.unwrap_or(false) {
                self.render_grid(opts.grid_size.unwrap_or(20));
            }
        }

        // 渲染所有元素
        for element in elements {
            if element.visible {
                self.render_element(element)?;
            }
        }

        Ok(())
    }

    /// 渲染单个元素
    fn render_element(&mut self, element: &RenderElement) -> Result<()> {
        self.surface.canvas().save();

        // 应用变换
        self.apply_transform(&element.transform);

        // 应用透明度
        if let Some(opacity) = element.style.opacity {
            if opacity < 1.0 {
                self.surface.canvas().save_layer_alpha(None, (opacity * 255.0) as u32);
            }
        }

        // 根据类型渲染
        match element.element_type {
            ElementType::Text => self.render_text(element)?,
            ElementType::Rect => self.render_rect(element)?,
            ElementType::Circle => self.render_circle(element)?,
            ElementType::Path => self.render_path(element)?,
            ElementType::Image => self.render_image(element)?,
            ElementType::Group => {
                if let Some(children) = &element.children {
                    for child in children {
                        self.render_element(child)?;
                    }
                }
            }
        }

        // 恢复透明度层
        if let Some(opacity) = element.style.opacity {
            if opacity < 1.0 {
                self.surface.canvas().restore();
            }
        }

        self.surface.canvas().restore();
        Ok(())
    }

    /// 渲染文本
    fn render_text(&mut self, element: &RenderElement) -> Result<()> {
        let data = &element.data;
        let content = data["content"].as_str().unwrap_or("");
        let font_size = data["fontSize"].as_f64().unwrap_or(14.0) as scalar;
        let font_family = data["fontFamily"].as_str().unwrap_or("Arial");
        let color = data["color"].as_str().unwrap_or("#000000");

        let mut paint = Paint::default();
        paint.set_anti_alias(true);
        paint.set_color(self.parse_color(color));

        // 创建字体
        let typeface = self.font_mgr
            .match_family_style(font_family, FontStyle::normal())
            .unwrap_or_else(|| self.font_mgr.match_family_style("", FontStyle::normal()).expect("Failed to get default typeface"));
        let font = Font::from_typeface(typeface, font_size);

        // 处理多行文本
        if content.contains('\n') {
            let lines: Vec<&str> = content.split('\n').collect();
            let mut y = font_size;
            for line in lines {
                self.surface.canvas().draw_str(line, SkPoint::new(0.0, y), &font, &paint);
                y += font_size * 1.2; // 行高
            }
        } else {
            self.surface.canvas().draw_str(content, SkPoint::new(0.0, font_size), &font, &paint);
        }

        Ok(())
    }

    /// 渲染矩形
    fn render_rect(&mut self, element: &RenderElement) -> Result<()> {
        let rect = Rect::from_xywh(
            0.0, 0.0,
            element.style.width.unwrap_or(100.0),
            element.style.height.unwrap_or(100.0),
        );

        let mut paint = Paint::default();
        paint.set_anti_alias(true);

        // 填充
        if let Some(fill) = &element.style.fill {
            paint.set_style(PaintStyle::Fill);
            paint.set_color(self.parse_color(fill));
            self.surface.canvas().draw_rect(rect, &paint);
        }

        // 描边
        if let Some(stroke) = &element.style.stroke {
            paint.set_style(PaintStyle::Stroke);
            paint.set_color(self.parse_color(stroke));
            paint.set_stroke_width(element.style.stroke_width.unwrap_or(1.0));
            self.surface.canvas().draw_rect(rect, &paint);
        }

        Ok(())
    }

    /// 渲染圆形
    fn render_circle(&mut self, element: &RenderElement) -> Result<()> {
        let radius = (element.style.width.unwrap_or(100.0).min(
            element.style.height.unwrap_or(100.0)
        )) / 2.0;

        let center = SkPoint::new(
            element.style.width.unwrap_or(100.0) / 2.0,
            element.style.height.unwrap_or(100.0) / 2.0,
        );

        let mut paint = Paint::default();
        paint.set_anti_alias(true);

        // 填充
        if let Some(fill) = &element.style.fill {
            paint.set_style(PaintStyle::Fill);
            paint.set_color(self.parse_color(fill));
            self.surface.canvas().draw_circle(center, radius, &paint);
        }

        // 描边
        if let Some(stroke) = &element.style.stroke {
            paint.set_style(PaintStyle::Stroke);
            paint.set_color(self.parse_color(stroke));
            paint.set_stroke_width(element.style.stroke_width.unwrap_or(1.0));
            self.surface.canvas().draw_circle(center, radius, &paint);
        }

        Ok(())
    }

    /// 渲染路径
    fn render_path(&mut self, element: &RenderElement) -> Result<()> {
        let mut path = Path::new();

        if let Some(commands) = element.data["commands"].as_array() {
            for cmd in commands {
                let cmd_type = cmd["type"].as_str().unwrap_or("");
                match cmd_type {
                    "M" => {
                        let x = cmd["x"].as_f64().unwrap_or(0.0) as scalar;
                        let y = cmd["y"].as_f64().unwrap_or(0.0) as scalar;
                        path.move_to((x, y));
                    }
                    "L" => {
                        let x = cmd["x"].as_f64().unwrap_or(0.0) as scalar;
                        let y = cmd["y"].as_f64().unwrap_or(0.0) as scalar;
                        path.line_to((x, y));
                    }
                    "Q" => {
                        let x1 = cmd["x1"].as_f64().unwrap_or(0.0) as scalar;
                        let y1 = cmd["y1"].as_f64().unwrap_or(0.0) as scalar;
                        let x = cmd["x"].as_f64().unwrap_or(0.0) as scalar;
                        let y = cmd["y"].as_f64().unwrap_or(0.0) as scalar;
                        path.quad_to((x1, y1), (x, y));
                    }
                    "C" => {
                        let x1 = cmd["x1"].as_f64().unwrap_or(0.0) as scalar;
                        let y1 = cmd["y1"].as_f64().unwrap_or(0.0) as scalar;
                        let x2 = cmd["x2"].as_f64().unwrap_or(0.0) as scalar;
                        let y2 = cmd["y2"].as_f64().unwrap_or(0.0) as scalar;
                        let x = cmd["x"].as_f64().unwrap_or(0.0) as scalar;
                        let y = cmd["y"].as_f64().unwrap_or(0.0) as scalar;
                        path.cubic_to((x1, y1), (x2, y2), (x, y));
                    }
                    "Z" => {
                        path.close();
                    }
                    _ => {}
                }
            }
        }

        let mut paint = Paint::default();
        paint.set_anti_alias(true);

        // 填充
        if let Some(fill) = &element.style.fill {
            paint.set_style(PaintStyle::Fill);
            paint.set_color(self.parse_color(fill));
            self.surface.canvas().draw_path(&path, &paint);
        }

        // 描边
        if let Some(stroke) = &element.style.stroke {
            paint.set_style(PaintStyle::Stroke);
            paint.set_color(self.parse_color(stroke));
            paint.set_stroke_width(element.style.stroke_width.unwrap_or(1.0));
            self.surface.canvas().draw_path(&path, &paint);
        }

        Ok(())
    }

    /// 渲染图片
    fn render_image(&mut self, element: &RenderElement) -> Result<()> {
        // 简化实现：暂时显示占位符
        let rect = Rect::from_xywh(
            0.0, 0.0,
            element.style.width.unwrap_or(100.0),
            element.style.height.unwrap_or(100.0),
        );

        let mut paint = Paint::default();
        paint.set_style(PaintStyle::Fill);
        paint.set_color(Color::from_rgb(240, 240, 240));
        self.surface.canvas().draw_rect(rect, &paint);

        // 绘制占位文字
        paint.set_color(Color::from_rgb(128, 128, 128));
        let default_typeface = self.font_mgr.match_family_style("", FontStyle::normal()).expect("Failed to get default typeface");
        let font = Font::from_typeface(default_typeface, 12.0);
        self.surface.canvas().draw_str(
            "[Image]",
            SkPoint::new(rect.width() / 2.0 - 20.0, rect.height() / 2.0),
            &font,
            &paint,
        );

        Ok(())
    }

    /// 渲染网格
    fn render_grid(&mut self, grid_size: i32) {
        let mut paint = Paint::default();
        paint.set_style(PaintStyle::Stroke);
        paint.set_color(Color::from_rgb(230, 230, 230));
        paint.set_stroke_width(0.5);
        paint.set_anti_alias(true);

        // 绘制垂直线
        for x in (0..self.width).step_by(grid_size as usize) {
            self.surface.canvas().draw_line(
                SkPoint::new(x as scalar, 0.0),
                SkPoint::new(x as scalar, self.height as scalar),
                &paint,
            );
        }

        // 绘制水平线
        for y in (0..self.height).step_by(grid_size as usize) {
            self.surface.canvas().draw_line(
                SkPoint::new(0.0, y as scalar),
                SkPoint::new(self.width as scalar, y as scalar),
                &paint,
            );
        }
    }

    /// 应用变换
    fn apply_transform(&mut self, transform: &Transform) {
        if let Some(translate) = &transform.translate {
            self.surface.canvas().translate((translate.x, translate.y));
        }
        if let Some(scale) = &transform.scale {
            self.surface.canvas().scale((scale.x, scale.y));
        }
        if let Some(rotate) = transform.rotate {
            if let Some(origin) = &transform.origin {
                self.surface.canvas().rotate(rotate, Some(SkPoint::new(origin.x, origin.y)));
            } else {
                self.surface.canvas().rotate(rotate, None);
            }
        }
    }

    /// 解析颜色
    fn parse_color(&self, color: &str) -> Color {
        if color.starts_with('#') {
            let hex = color.trim_start_matches('#');
            let r = u8::from_str_radix(&hex[0..2], 16).unwrap_or(0);
            let g = u8::from_str_radix(&hex[2..4], 16).unwrap_or(0);
            let b = u8::from_str_radix(&hex[4..6], 16).unwrap_or(0);
            let a = if hex.len() > 6 {
                u8::from_str_radix(&hex[6..8], 16).unwrap_or(255)
            } else {
                255
            };
            Color::from_argb(a, r, g, b)
        } else {
            Color::BLACK
        }
    }

    // === 导出功能 ===

    /// 导出为 PNG
    pub fn export_png(&mut self, elements: &[RenderElement]) -> Result<Vec<u8>> {
        self.render(elements)?;

        let image = self.surface.image_snapshot();
        let data = image.encode_to_data(EncodedImageFormat::PNG)
            .ok_or_else(|| anyhow!("Failed to encode PNG"))?;

        Ok(data.as_bytes().to_vec())
    }

    /// 导出为 JPEG
    pub fn export_jpg(&mut self, elements: &[RenderElement], quality: u32) -> Result<Vec<u8>> {
        self.render(elements)?;

        let image = self.surface.image_snapshot();
        let data = image.encode_to_data_with_quality(EncodedImageFormat::JPEG, quality)
            .ok_or_else(|| anyhow!("Failed to encode JPEG"))?;

        Ok(data.as_bytes().to_vec())
    }

    /// 导出为 PDF
    pub fn export_pdf(&mut self, _elements: &[RenderElement]) -> Result<Vec<u8>> {
        // TODO: Fix PDF export - API has changed in current skia-safe version
        // Need to update the implementation for the new API
        // let mut document = pdf::new_document(None, None);
        // let page_size = (self.width as scalar, self.height as scalar);
        // let canvas = document.begin_page(page_size, None);
        // for element in elements {
        //     if element.visible {
        //         self.render_element(&mut canvas, element)?;
        //     }
        // }
        // document.end_page();
        // let data = document.close();
        // Ok(data.as_bytes().to_vec())
        Ok(Vec::new())
    }

    /// 导出为 SVG
    pub fn export_svg(&mut self, _elements: &[RenderElement]) -> Result<String> {
        // TODO: Fix SVG export - svg::Canvas doesn't support mutable borrow for render_element
        // Need to implement a different approach for SVG export
        // let bounds = Rect::from_wh(self.width as scalar, self.height as scalar);
        // let mut svg_canvas = svg::Canvas::new(bounds, None);
        // for element in elements {
        //     if element.visible {
        //         self.render_element(&mut svg_canvas, element)?;
        //     }
        // }
        // Ok(svg_canvas.end())
        Ok(String::new())
    }

    /// 导出为 WebP
    pub fn export_webp(&mut self, elements: &[RenderElement], quality: u32) -> Result<Vec<u8>> {
        self.render(elements)?;

        let image = self.surface.image_snapshot();
        let data = image.encode_to_data_with_quality(EncodedImageFormat::WEBP, quality)
            .ok_or_else(|| anyhow!("Failed to encode WebP"))?;

        Ok(data.as_bytes().to_vec())
    }
}